src/components/Context/CartContext.tsx

```diff
@@ -0,0 +1,100 @@
+import { PropsWithChildren, createContext, useState } from "react";
+import { FeaturedProductObject } from "../App";
+
+export interface ItemToPurchase {
+  quantity: number;
+  product: FeaturedProductObject;
+}
+
+export const CartContext = createContext<ItemToPurchase[]>([]);
+
+export const CartProvider = ({ children }: PropsWithChildren) => {
+  const [itemsInCart, setItemsInCart] = useState<ItemToPurchase[]>([]);
+
+  const numItemsInCart = () => {
+    return itemsInCart.length;
+  };
+
+  const addItemToCart = (itemToPurchase: ItemToPurchase) => {
+    const isItemInCart = itemsInCart.find((itemInCart) => {
+      return itemToPurchase.product.id === itemInCart.product.id; // will find item if it's already in the cart, or else will return undefined
+    });
+    if (isItemInCart) {
+      setItemsInCart(
+        itemsInCart.map((itemInCart) => {
+          return itemInCart.product.id === itemToPurchase.product.id
+            ? {
+                ...itemInCart,
+                quantity: itemInCart.quantity + itemToPurchase.quantity,
+              }
+            : itemInCart;
+        })
+      );
+    } else {
+      setItemsInCart([...itemsInCart, itemToPurchase]);
+    }
+  };
+
+  const removeItemFromCart = (itemToRemove: ItemToPurchase) => {
+    setItemsInCart(
+      itemsInCart.map((itemInCart) => {
+        return itemInCart.product.id === itemToRemove.product.id
+          ? { ...itemInCart, quantity: (itemInCart.quantity -= 1) }
+          : itemInCart;
+      })
+    );
+
+    setItemsInCart(
```

Comment:
it would be slightly better to avoid 2 calls to `setItemsInCart`, otherwise you could get flickers in the UI. also, it's important to avoid mutating any objects like on line 42. I'd suggest this, almost exactly what you have but just with those 2 differences:

```ts
const newItems = itemsInCart.map(item => item.product.id === itemToRemove.prodcut.id
  ? { ...item, quantity: item.quantity - 1 })
  : item
).filter(item => item.quantity > 0);
setItemsInCart(newItems);
```

I remember you said something about needing a `-=`, was that here? you shouldn't need it, lmk if you start seeing problems again if you decide to try this snippet

Reply:
This is exactly where I was running into that error!  I'll let you know if it happens again and will make this fix, too!!